### ISSUE TITLE
Undef + Seek on filehandles opened to references causes segfaults

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8617]
Migrated from [rt.perl.org#40407](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=40407) (status was 'resolved')

Searchable as RT40407$

**Comment by p5pRT** [comment]
### From buu@erxz.com

#### Created by buu@erxz.com



The following snippet segfaults for me&#8203;:

my $scalar1;
open my $fh\, ">"\, \\$scalar1 or die $\!;

for\( 1\.\.2 \)
\{
&nbsp;    undef $scalar1;
&nbsp;    seek\($fh\,0\,0\);
&nbsp;    print $fh "a";
\}


It also seems to segfault under other versions of perl&#8203;:  
5\.8\.6/darwin\, 5\.8\.8/FreeBSD and 5\.9\.4/FreeBSD

I'm not sure this is completely useful or even rational code\, but it probably shouldn't segfault\.

Also as a secondary note\, truncate on filehandles opened to references fails to do anything useful\,
but it at least fails gracefully\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.8:

Configured by Debian Project at Sun Aug  6 15:47:28 UTC 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.15-1-686, archname=i486-linux-gnu-thread-multi
    uname='linux ulises 2.6.15-1-686 #2 mon mar 6 15:27:08 utc 2006 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.8 -Dsitearch=/usr/local/lib/perl/5.8.8 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.8 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.1.2 20060729 (prerelease) (Debian 4.1.1-10)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.6.so, so=so, useshrplib=true, libperl=libperl.so.5.8.8
    gnulibc_version='2.3.6'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /etc/perl
    /usr/local/lib/perl/5.8.8
    /usr/local/share/perl/5.8.8
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    /usr/local/lib/perl/5.8.7
    /usr/local/share/perl/5.8.7
    /usr/local/lib/perl/5.8.4
    /usr/local/share/perl/5.8.4
    .


Environment for perl v5.8.8:
    HOME=/home/buu
    LANG=en_US
    LANGUAGE=en_US:en_GB:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/buu/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

This bug has been fixed by the following change&#8203;:

Change 29751 by rgs@&#8203;stencil on 2007/01/11 10&#8203;:34&#8203;:02

&nbsp;        Fix bug \#40407&#8203;: after a seek on a PerlIO&#8203;::scalar filehandle\,
&nbsp;        ensure there's a string buffer in the scalar

Affected files \.\.\.

\.\.\. //depot/perl/ext/PerlIO/scalar/scalar\.xs\#16 edit

Differences \.\.\.

==== //depot/perl/ext/PerlIO/scalar/scalar\.xs\#16 \(text\) ====

@&#8203;@&#8203; \-108\,6 \+108\,10 @&#8203;@&#8203;
&nbsp;        Zero\(SvPVX\(s\->var\) \+ oldcur\, newlen \- oldcur\, char\);
&nbsp;        /\* No SvCUR\_set\(\)\, though\.  This is just a seek\, not a write\. \*/
&nbsp;     \}
\+    else if \(\!SvPVX\(s\->var\)\) \{
\+       /\* ensure there's always a character buffer \*/
\+       \(void\)SvGROW\(s\->var\,1\);
\+    \}
&nbsp;     SvPOK\_on\(s\->var\);
&nbsp;     return 0;
 \}

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

This bug has been fixed by the following change&#8203;:

Change 29751 by rgs@&#8203;stencil on 2007/01/11 10&#8203;:34&#8203;:02

&nbsp;        Fix bug \#40407&#8203;: after a seek on a PerlIO&#8203;::scalar filehandle\,
&nbsp;        ensure there's a string buffer in the scalar

Affected files \.\.\.

\.\.\. //depot/perl/ext/PerlIO/scalar/scalar\.xs\#16 edit

Differences \.\.\.

==== //depot/perl/ext/PerlIO/scalar/scalar\.xs\#16 \(text\) ====

@&#8203;@&#8203; \-108\,6 \+108\,10 @&#8203;@&#8203;
&nbsp;        Zero\(SvPVX\(s\->var\) \+ oldcur\, newlen \- oldcur\, char\);
&nbsp;        /\* No SvCUR\_set\(\)\, though\.  This is just a seek\, not a write\. \*/
&nbsp;     \}
\+    else if \(\!SvPVX\(s\->var\)\) \{
\+       /\* ensure there's always a character buffer \*/
\+       \(void\)SvGROW\(s\->var\,1\);
\+    \}
&nbsp;     SvPOK\_on\(s\->var\);
&nbsp;     return 0;
 \}

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
